src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/ICustomMarshalerTests.cs

```diff
@@ -0,0 +1,537 @@
+        {
+            Assert.Throws<ApplicationException>(() => NonICustomMarshalerMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_OpenGenericICustomMarshaler_ThrowsTypeLoadException()
+        {
+            Assert.Throws<TypeLoadException>(() => OpenGenericICustomMarshalerMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_GetInstanceMethodDoesntExist_ThrowsApplicationException()
+        {
+            Assert.Throws<ApplicationException>(() => NoGetInstanceMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_GetInstanceMethodInstanceMethod_ThrowsApplicationException()
+        {
+            Assert.Throws<ApplicationException>(() => InstanceGetInstanceMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_GetInstanceMethodNoParameters_ThrowsApplicationException()
+        {
+            Assert.Throws<ApplicationException>(() => NoParametersGetInstanceMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_GetInstanceMethodNonStringParameter_ThrowsApplicationException()
+        {
+            Assert.Throws<ApplicationException>(() => NonStringGetInstanceMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_GetInstanceMethodReturnsVoid_ThrowsApplicationException()
+        {
+            Assert.Throws<ApplicationException>(() => VoidGetInstanceMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_GetInstanceMethodReturnsNull_ThrowsApplicationException()
+        {
+            Assert.Throws<ApplicationException>(() => NullGetInstanceMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_GetInstanceMethodThrows_ThrowsActualException()
+        {            
+            Assert.Throws<NotImplementedException>(() => ThrowingGetInstanceMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_MarshalManagedToNativeThrows_ThrowsActualException()
+        {
+            Assert.Throws<NotImplementedException>(() => ThrowingMarshalManagedToNativeMethod(""));
+        }
+
+        [Fact]
+        public void Parameter_CleanUpNativeDataMethodThrows_ThrowsActualException()
+        {
+            Assert.Throws<NotImplementedException>(() => ThrowingCleanUpNativeDataMethod(""));
+        }
+
+        [Fact]
+        public void Field_ParentIsStruct_ThrowsTypeLoadException()
+        {
+            Assert.Throws<TypeLoadException>(() => StructWithCustomMarshalerFieldMethod(new StructWithCustomMarshalerField()));
+        }
+
+        public class StringForwardingCustomMarshaler : ICustomMarshaler
+        {
+            public void CleanUpManagedData(object ManagedObj) { }
+            public void CleanUpNativeData(IntPtr pNativeData) { Marshal.ZeroFreeGlobalAllocAnsi(pNativeData); }
+
+            public int GetNativeDataSize() => -1;
+
+            public IntPtr MarshalManagedToNative(object ManagedObj) => Marshal.StringToCoTaskMemAnsi((string)ManagedObj);
```

Comment:
Your allocation and freeing need to match: you're allocating with CoTaskMem but freeing with GlobalAlloc.